daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,217 @@
+
+  /**
+   * Method to be called when a regular element is encountered.  This is one of the most
+   * important methods; most DFDL elements with useful data become some subclass of
+   * ElementBase.
+   * @param element the element that was encountered
+   */
+  protected def onElementBegin(element: ElementBase): Unit
+  /**
+   * Method to be called when a regular element has been processed.
+   * @param element the element that was encountered
+   */
+  protected def onElementEnd(element: ElementBase): Unit
+
+  final def walkDSOMSchema(): T = {
+    onWalkBegin(schemaSetRoot)
+    // this is allowed because Root is a subclass of ElementBase
+    walkerHelper(schemaSetRoot)
+    // onWalkEnd will return the value returned by this method!
+    onWalkEnd(schemaSetRoot)
+  }
+
+  /**
+   * Overloaded walkerHelper method used for when some "group" element is encountered in
+   * the DSOM; the group elements are Sequences, Choices, and GroupRefs
+   * @param modelGroup the ModelGroup element that was encountered in the Schema
+   */
+  private final def walkerHelper(modelGroup: ModelGroup): Unit = {
+    modelGroup match {
+      case sequence: Sequence =>
+        onSequenceBegin(sequence)
+        for (child <- sequence.groupMembers) walkerHelper(child)
+        onSequenceEnd(sequence)
+      case choice: Choice =>
+        onChoiceBegin(choice)
+        for (child <- choice.groupMembers) walkerHelper(child)
+        onChoiceEnd(choice)
+      case groupRef: GroupRef =>
+        if (!groupRef.isHidden) {
+          onGroupRefBegin(groupRef)
+          for (child <- groupRef.groupMembers) walkerHelper(child)
+          onGroupRefEnd(groupRef)
+        }
+    }
+  }
+
+  /**
+   * Overloaded walkerHelper method that will handle the input either as an ElementBase or ModelGroup
+   * @param term the Term element that was encountered in the Schema
+   */
+  private final def walkerHelper(term: Term): Unit = {
+    term match {
+      case element: ElementBase => walkerHelper(element)
+      case modelGroup: ModelGroup => walkerHelper(modelGroup)
+    }
+  }
+
+  /**
+   * Overloaded walkerHelper method that will handle some element either as a complex or simple type
+   * @param element the element that was encountered in the Schema
+   */
+  private final def walkerHelper(element: ElementBase): Unit = {
+    onElementBegin(element)
+    if (element.isComplexType) {
+      onComplexBegin(element.complexType)
+      walkerHelper(element.complexType.group)
+      onComplexEnd(element.complexType)
+    } else {
+      onSimpleBegin(element.simpleType)
+      onSimpleEnd(element.simpleType)
+    }
+    onElementEnd(element)
+  }
+}
+
+/**
+ * The sole purpose of this Companion Object is to contain this one static variable.
+ */
+object AbstractDSOMWalker {
+  /**
+   * Used to suppress all of the extra printing from running test cases
+   */
+  final val PRODUCTION_MODE: Boolean = true
+}
```

Comment:
This is only used in the NiFI specific Record stuff. I would expect this to just be a flag in the RecordWalker. Presumable it would plug into NiFi's logging mechanism.

Reply:
I agree, that's the more appropriate location.  I might get rid of this flag eventually, the only reason it exists is if the Unit tests fail you can turn it on and see what the data structures look like at each stage of the processing.